File: algorithm/perfect_squares/perfect_squares.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>

using std::vector;
using std::sqrt;
using std::min;

class Solution {
public:
	int numSquares(int n) {
		vector<int> count(n+1, n+1);
		count[0] = 0;
		count[1] = 1;
		for(int i = 2; i <= n; i++){
			int a = 1;
			while(a*a <= i) {
				count[i] = min(count[i], count[i-a*a]+1);
				a++;
			}
		}
		return count[n];
	}
};

File: algorithm/perfect_squares/go/perfect_squares.go
package main

import (
	"fmt"
)

func numSquares(n int) int {
	count := make([]int, 1, n+1)
	for len(count) <= n {
		m := len(count)
		currentMin := m
		for i := 1; i*i <= m; i++ {
			currentMin = min(currentMin, count[m-i*i]+1)
		}
		count = append(count, currentMin)
	}
	return count[n]
}

func min(x, y int) int {
	if x > y {
		return y
	}
	return x
}

func main() {
	num := 100
	for i := 1; i < num; i++ {
		fmt.Println(i, "has been composed by", numSquares(i))
	}
}
